### PR TITLE
Add preference to en/disable Shared pinned tab

### PR DESCRIPTION
**Pull request by sangwoo108**
Resolves https://github.com/brave/brave-browser/issues/38509

## Submitter Checklist:

- [ ] I confirm that no [security/privacy review is needed](https://github.com/brave/brave-browser/wiki/Security-reviews) and no other type of reviews are needed, or that I have [requested](https://github.com/brave/reviews/issues/new/choose) them
- [ ] There is a [ticket](https://github.com/brave/brave-browser/issues) for my issue
- [ ] Used Github [auto-closing keywords](https://docs.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue) in the PR description above
- [ ] Wrote a good [PR/commit description](https://google.github.io/eng-practices/review/developer/cl-descriptions.html)
- [ ] Squashed any review feedback or "fixup" commits before merge, so that history is a record of what happened in the repo, not your PR
- [ ] Added appropriate labels (`QA/Yes` or `QA/No`; `release-notes/include` or `release-notes/exclude`; `OS/...`) to the associated issue
- [ ] Checked the PR locally:
  * `npm run test -- brave_browser_tests`, `npm run test -- brave_unit_tests` [wiki](https://github.com/brave/brave-browser/wiki/Tests)
  * `npm run presubmit` [wiki](https://github.com/brave/brave-browser/wiki/Presubmit-checks), `npm run gn_check`, `npm run tslint`
- [ ] Ran `git rebase master` (if needed)

## Reviewer Checklist:

- [ ] A security review [is not needed](https://github.com/brave/brave-browser/wiki/Security-reviews), or a link to one is included in the PR description
- [ ] New files have MPL-2.0 license header
- [ ] Adequate test coverage exists to prevent regressions
- [ ] Major classes, functions and non-trivial code blocks are well-commented
- [ ] Changes in component dependencies are properly reflected in `gn`
- [ ] Code follows the [style guide](https://chromium.googlesource.com/chromium/src/+/HEAD/styleguide/c++/c++.md)
- [ ] Test plan is specified in PR before merging

## After-merge Checklist:

- [ ] The associated issue milestone is set to the smallest version that the
  changes has landed on
- [ ] All relevant documentation has been updated, for instance:
  - [ ] https://github.com/brave/brave-browser/wiki/Deviations-from-Chromium-(features-we-disable-or-remove)
  - [ ] https://github.com/brave/brave-browser/wiki/Proxy-redirected-URLs
  - [ ] https://github.com/brave/brave-browser/wiki/Fingerprinting-Protections
  - [ ] https://github.com/brave/brave-browser/wiki/Brave%E2%80%99s-Use-of-Referral-Codes
  - [ ] https://github.com/brave/brave-browser/wiki/Web-Compatibility-Exceptions-in-Brave
  - [ ] https://github.com/brave/brave-browser/wiki/QA-Guide
  - [ ] https://github.com/brave/brave-browser/wiki/P3A

### CODE REVIEW

File: browser/ui/toolbar/brave_location_bar_model_delegate.h [hunk 1]
```diff
@@ -26,6 +26,8 @@ class BraveLocationBarModelDelegate : public BrowserLocationBarModelDelegate {
       const GURL& url,
       const std::u16string& formatted_url) const override;
   bool GetURL(GURL* url) const override;
+
+  raw_ptr<Browser> browser_;
```
Comment (simonhong) [inline browser/ui/toolbar/brave_location_bar_model_delegate.h:30]: nit: `= nullptr` or `raw_ref`?

File: browser/ui/tabs/shared_pinned_tab_service.cc [hunk 2]
```diff
@@ -196,7 +198,29 @@ SharedPinnedTabService::SharedPinnedTabService(Profile* profile)
...
   profile_observation_.Observe(profile_);
-  browser_list_observation_.Observe(BrowserList::GetInstance());
+
+  shared_pinned_tab_enabled_.Init(
+      brave_tabs::kSharedPinnedTab, profile_->GetPrefs(),
+      base::BindRepeating(
+          [](base::WeakPtr<SharedPinnedTabService> service) {
+            if (UNLIKELY(!service)) {
+              // As BooleanPrefMember is already RAII style subscription,
+              // this is unlikely to happen but theoretically might happen on
```
Comment (simonhong) [inline browser/ui/tabs/shared_pinned_tab_service.cc:208]: Why do you think it's available in theory?
As `shared_pinned_tab_enabled_` is member var of this service, it'll be destroyed before service is gone.
Also `weak_ptr_factory_` is last declared member, so it should not be null when it's referred by other member vars.

**Review by simonhong — APPROVED** [review]
++ 👍🏼

File: browser/ui/tabs/shared_pinned_tab_service.cc (hunk 2 above)
Comment (sangwoo108) [reply browser/ui/tabs/shared_pinned_tab_service.cc:208]: Oh wait, shared_pinned_tab_enabled_ can be alive while weak_ptr_factory_ is invalidated, as destruction order is from the last? I think making a method would be clear.

Comment (simonhong) [reply browser/ui/tabs/shared_pinned_tab_service.cc]: Ah, right. it's invalidated at first than all others.

File: app/brave_settings_strings.grdp [hunk 3]
```diff
@@ -113,6 +113,9 @@
   <message name="IDS_SETTINGS_APPEARANCE_SETTINGS_BRAVE_TAB_HOVER_MODE_TOOLTIP" desc="The label name of the tab hover mode where a tooltip with the site title is shown.">
     Tooltip
   </message>
+  <message name="IDS_SETTINGS_APPEARANCE_SETTINGS_BRAVE_SHARED_PINNED_TAB" desc="The label name of a toggle button for Shared Pinned Tab feature">
+    Shared pinned tabs
```
Comment (mkarolin) [inline app/brave_settings_strings.grdp:117]: @sangwoo108 can we be a little more descriptive here as to what this toggle would do? For example, `Show pinned tabs in all windows`. cc: @rebron do we have a better wording for this feature? "Shared pinned tabs" as a settings toggle is not immediately obvious, imo.

Comment (sangwoo108) [reply app/brave_settings_strings.grdp:117]: That's much better!

**Review by mkarolin — APPROVED** [review]
`strings`++

**Pull request merged**